File: sdks/hsafa-sdk/src/schema.ts
// =============================================================================
// @hsafa/sdk — Schema Converter
// Converts simple "string" / "number?" / "string[]" type strings to JSON Schema
// =============================================================================

export function inputToJsonSchema(input: Record<string, string>): Record<string, unknown> {
  const properties: Record<string, unknown> = {};
  const required: string[] = [];

  for (const [key, typeStr] of Object.entries(input)) {
    const optional = typeStr.endsWith('?');
    const baseType = optional ? typeStr.slice(0, -1) : typeStr;

    if (!optional) required.push(key);

    if (baseType === 'string[]') {
      properties[key] = { type: 'array', items: { type: 'string' } };
    } else if (baseType === 'number[]') {
      properties[key] = { type: 'array', items: { type: 'number' } };
    } else if (baseType === 'boolean[]') {
      properties[key] = { type: 'array', items: { type: 'boolean' } };
    } else if (baseType === 'object') {
      properties[key] = { type: 'object', additionalProperties: true };
    } else {
      properties[key] = { type: baseType };
    }
  }

  const schema: Record<string, unknown> = {
    type: 'object',
    properties,
    additionalProperties: false,
  };

  if (required.length > 0) schema.required = required;

  return schema;
}

// Best-effort partial JSON parsing — accumulates text and tries to parse on each delta
export function parsePartialJson(accumulated: string): Record<string, unknown> {
  // Try direct parse first
  try {
    return JSON.parse(accumulated) as Record<string, unknown>;
  } catch {
    // Try adding closing braces
    const attempts = [
      accumulated + '}',
      accumulated + '"}',
      accumulated + '"}]',
      accumulated + '"}]}',
    ];
    for (const attempt of attempts) {
      try {
        return JSON.parse(attempt) as Record<string, unknown>;
      } catch {
        // continue
      }
    }
    return {};
  }
}

File: sdks/python/src/hsafa_sdk/schema.py
import json
from typing import Dict, Any


def input_to_json_schema(input_types: Dict[str, str]) -> Dict[str, Any]:
    properties: Dict[str, Any] = {}
    required: list[str] = []

    for key, type_str in input_types.items():
        optional = type_str.endswith('?')
        base_type = type_str[:-1] if optional else type_str

        if not optional:
            required.append(key)

        if base_type == 'string[]':
            properties[key] = {'type': 'array', 'items': {'type': 'string'}}
        elif base_type == 'number[]':
            properties[key] = {'type': 'array', 'items': {'type': 'number'}}
        elif base_type == 'boolean[]':
            properties[key] = {'type': 'array', 'items': {'type': 'boolean'}}
        elif base_type == 'object':
            properties[key] = {'type': 'object', 'additionalProperties': True}
        else:
            properties[key] = {'type': base_type}

    schema: Dict[str, Any] = {
        'type': 'object',
        'properties': properties,
        'additionalProperties': False,
    }

    if required:
        schema['required'] = required

    return schema


def parse_partial_json(accumulated: str) -> Dict[str, Any]:
    try:
        return json.loads(accumulated)
    except json.JSONDecodeError:
        attempts = [
            accumulated + '}',
            accumulated + '"}',
            accumulated + '"}]',
            accumulated + '"}]}',
        ]
        for attempt in attempts:
            try:
                return json.loads(attempt)
            except json.JSONDecodeError:
                continue
        return {}
